applications/FluidDynamicsApplication/custom_processes/distance_smoothing_process.cpp

```diff
@@ -0,0 +1,228 @@
+#include "distance_smoothing_process.h"
+
+
+namespace Kratos
+{
+
+/* Public functions *******************************************************/
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver>
+DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::DistanceSmoothingProcess(
+    ModelPart& rModelPart,
+    typename TLinearSolver::Pointer plinear_solver)
+    : Process(),
+      mrModelPart(rModelPart)
+{
+    // Generate an auxilary model part and populate it by elements of type DistanceSmoothingElement
+    CreateAuxModelPart();
+
+    auto p_builder_solver = Kratos::make_shared<ResidualBasedBlockBuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver> >(plinear_solver);
+
+    CreateSolutionStrategy(plinear_solver, p_builder_solver);
+}
+
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver>
+DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::DistanceSmoothingProcess(
+    ModelPart& rModelPart,
+    Parameters& rParameters)
+    : Process(),
+      mrModelPart(rModelPart)
+{
+    // Generate an auxilary model part and populate it by elements of type DistanceSmoothingElement
+    CreateAuxModelPart();
+
+    typename TLinearSolver::Pointer plinear_solver = LinearSolverFactory<TSparseSpace, TDenseSpace>().Create(
+        rParameters["linear_solver_settings"]);
+
+    auto p_builder_solver = Kratos::make_shared<ResidualBasedBlockBuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver> >(plinear_solver);
+
+    CreateSolutionStrategy(plinear_solver, p_builder_solver);
+}
+
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver>
+DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::DistanceSmoothingProcess(
+    Model &rModel,
+    Parameters &rParameters)
+    : Process(),
+      mrModelPart(rModel.GetModelPart(rParameters["model_part_name"].GetString()))
+{
+    // Generate an auxilary model part and populate it by elements of type DistanceSmoothingElement
+    CreateAuxModelPart();
+
+    typename TLinearSolver::Pointer plinear_solver = LinearSolverFactory<TSparseSpace, TDenseSpace>().Create(
+        rParameters["linear_solver_settings"]);
+
+    auto p_builder_solver = Kratos::make_shared<ResidualBasedBlockBuilderAndSolver<TSparseSpace, TDenseSpace, TLinearSolver> >(plinear_solver);
+
+    CreateSolutionStrategy(plinear_solver, p_builder_solver);
+}
+
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver>
+void DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::CreateAuxModelPart()
+{
+    Model& current_model = mrModelPart.GetModel();
+    if(current_model.HasModelPart( mAuxModelPartName ))
+        current_model.DeleteModelPart( mAuxModelPartName );
+
+    // Generate AuxModelPart
+    ModelPart& r_smoothing_model_part = current_model.CreateModelPart( mAuxModelPartName );
```

Comment:
please make this name somehow dependent on the name of the original ModelPart, otherwise you can only create one instance of this process!

see [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp#L68)

Reply:
> please make this name somehow dependent on the name of the original ModelPart, otherwise you can only create one instance of this process!
> 
> see [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp#L68)

@philbucher 
Now, in the destruction, current_model.DeleteModelPart returns Segmentation fault!
Could you please take a look at it
https://github.com/KratosMultiphysics/Kratos/blob/9c425ee7c570cd37b6f019ac9fb300d68c4b96e0/applications/FluidDynamicsApplication/custom_processes/distance_smoothing_process.h#L91

It is strange since the if condition is passed. Also, previously with an arbitrary model_part name this doesn't happen.